### PR TITLE
[http-client-csharp] fix: gen. unknown discriminator model when discriminator prop exists

### PR DESCRIPTION
**Pull request by jorgerangel-msft**
fixes: https://github.com/microsoft/typespec/issues/4935

### CODE REVIEW

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeDefinition.cs [hunk 1]
```diff
@@ -442,13 +442,21 @@ internal MethodProvider BuildDeserializationMethod()
         private MethodBodyStatement[] BuildAbstractDeserializationMethodBody()
         {
             var unknownVariant = _model.DerivedModels.First(m => m.IsUnknownDiscriminatorModel);
+            bool onlyContainsUnknownDerivedModel = _model.DerivedModels.Count == 1;
+            IfStatement? deserializeDiscriminatedModelsConditions = null;
+
+            if (!onlyContainsUnknownDerivedModel)
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeDefinition.cs:448]: nit: we can just always initialize `deserializeDiscriminatedModelsCondition` and then check for `onlyContainsUnknownDerivedModel` on L459.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.Input/src/InputTypes/InputModelType.cs [hunk 2]
```diff
@@ -79,12 +80,15 @@ public IReadOnlyDictionary<string, InputModelType> DiscriminatedSubtypes
...
             {
-                if (value is null || value.Count == 0)
+                if (value is null || DiscriminatorProperty == null)
                     return;
 
                 _discriminatedSubtypes = new Dictionary<string, InputModelType>(value);
-                var cleanBaseName = Name.ToCleanName();
-                _discriminatedSubtypes.Add("unknown",
+
+                if (DiscriminatorValue != UnknownDiscriminatorValue)
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.Input/src/InputTypes/InputModelType.cs:88]: Do we need to support adding an "Unknown" discriminator in the tsp? This has a semantically special meaning to our generator where it is the fallback, but if it is defined in the tsp, it seems like they would want to match on the literal string "Unknown". I think we can probably leave this off for now until we here of a use case for it.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.Input/src/InputTypes/Serialization/TypeSpecInputModelTypeConverter.cs [hunk 3]
```diff
@@ -105,10 +105,14 @@ public static InputModelType CreateModelType(ref Utf8JsonReader reader, string?
             {
                 model.Properties = properties;
             }
-            if (discriminatedSubtypes != null)
+            if (discriminatedSubtypes?.Count > 0)
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.Input/src/InputTypes/Serialization/TypeSpecInputModelTypeConverter.cs:108]: Why did we need to change this check?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.Input/src/InputTypes/InputModelType.cs (hunk 2 above)
Comment (jorgerangel-msft) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.Input/src/InputTypes/InputModelType.cs:88]: Yes this makes sense and I agree that it is a valid scenario even if it's unlikely.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.Input/src/InputTypes/Serialization/TypeSpecInputModelTypeConverter.cs (hunk 3 above)
Comment (jorgerangel-msft) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.Input/src/InputTypes/Serialization/TypeSpecInputModelTypeConverter.cs]: ah, this is remnants of another approach I initially took. I have reverted this.

**Review by JoshLove-msft — APPROVED** [review]

**Pull request merged**